### PR TITLE
[model_io] Fix the color parsing in the model_io

### PR DESCRIPTION
**Pull request by GiulioRomualdi**
Once `VisualElement::childElementForName` is called with input equal to 'material' a pointer to the
material info is stored inside the VisualElement.

## Rationale
Differently from [`GeometrcElement`](https://github.com/robotology/idyntree/blob/cbb0bf5e593b7add489bccb7dd3ba5054b200573/src/model_io/urdf/src/GeometryElement.cpp#L23), the constructor of `MaterialElement` takes as input a copy of `std::shared_ptr<MaterialInfo>` and not a **reference** to a shared pointer

https://github.com/robotology/idyntree/blob/cbb0bf5e593b7add489bccb7dd3ba5054b200573/src/model_io/urdf/src/MaterialElement.cpp#L23-L30

For this reason in the following line

https://github.com/robotology/idyntree/blob/cbb0bf5e593b7add489bccb7dd3ba5054b200573/src/model_io/urdf/src/VisualElement.cpp#L51

Since `m_info.m_material = nullptr` and `MaterialElement` copies the value, `m_info.m_material` [was not updated by 
](https://github.com/robotology/idyntree/blob/cbb0bf5e593b7add489bccb7dd3ba5054b200573/src/model_io/urdf/src/MaterialElement.cpp#L23-L30)
```cpp
     if (!m_info) { 
         m_info = std::make_shared<MaterialInfo>(); 
     } 
```

Passing a reference to a shared ptr cannot solve the problem in this particular case. Indeed given the following code: 

https://github.com/robotology/idyntree/blob/cbb0bf5e593b7add489bccb7dd3ba5054b200573/src/model_io/urdf/src/RobotElement.cpp#L53-L56

the compiler will complain since `nullptr` is a `r-value` and it cannot be converted into a `l-value`.

## Outcome of the PR

This is how Talos is now displayed with 

### meshcat

image

### irrlicht 

image

For iCub there is the usual problem with the colors 

### meshcat
image

### irrlicht

image

### CODE REVIEW

**Comment by traversaro** [comment]
Great, quite a clean fix. Can you add a line on the CHANGELOG?

If you plan to work on https://github.com/robotology/icub-models-generator/issues/216 we can wait for that also to be ready and merge/release both things together, otherwise no problem, we can merge this and enjoy a colorful iCub!

**Comment by traversaro** [comment]
By the way, checking the CI failure, the bugfix revealed a bug in the test models that was "hidden" by this bug:
~~~
Error:  URDFDocument :: addVisualPropertiesToModel : Material for link head has not rgba and it is not in the global material database.
Error:  URDFDocument :: documentHasBeenParsed : Failed to add visual elements to model
Model loaded from /home/runner/work/idyntree/idyntree/src/tests/data/iCubGenova02.urdf
~~~

Can you fix the vendored `/home/runner/work/idyntree/idyntree/src/tests/data/iCubGenova02.urdf` model? We could also try to fix it with a latest model from `icub-models`,but I am not sure if then we need to also fix other tests that have hardcoded expectations on the number of elements (however, we can try).

**Comment by traversaro** [comment]
> By the way, checking the CI failure, the bugfix revealed a bug in the test models that was "hidden" by this bug:
> 
> ```
> Error:  URDFDocument :: addVisualPropertiesToModel : Material for link head has not rgba and it is not in the global material database.
> Error:  URDFDocument :: documentHasBeenParsed : Failed to add visual elements to model
> Model loaded from /home/runner/work/idyntree/idyntree/src/tests/data/iCubGenova02.urdf
> ```
> 
> Can you fix the vendored `/home/runner/work/idyntree/idyntree/src/tests/data/iCubGenova02.urdf` model? We could also try to fix it with a latest model from `icub-models`,but I am not sure if then we need to also fix other tests that have hardcoded expectations on the number of elements (however, we can try).

I looked into the `iCubGenova02.urdf` model and actually it is a corner case that perhaps it is not clear from the URDF spec (i.e. http://wiki.ros.org/urdf/XML/link). The head link refer to material `head` (see https://github.com/robotology/idyntree/blob/cbb0bf5e593b7add489bccb7dd3ba5054b200573/src/tests/data/iCubGenova02.urdf#L1394), that is defined by the material element under the link `r_foot`  (see https://github.com/robotology/idyntree/blob/cbb0bf5e593b7add489bccb7dd3ba5054b200573/src/tests/data/iCubGenova02.urdf#L1394), and not by a material element under the robot element. An easy fix could be to either add all link materials to the database, or jus fix the `/home/runner/work/idyntree/idyntree/src/tests/data/iCubGenova02.urdf` model.

**Comment by GiulioRomualdi** [comment]
Hi, @traversaro I think it's easier to fix the model. Should update the model to the latest one stored in icub-models?

**Comment by traversaro** [comment]
> Should update the model to the latest one stored in icub-models?

You can try.

**Comment by GiulioRomualdi** [comment]
> > Should update the model to the latest one stored in icub-models?
> 
> You can try.

After a short discussion with @traversaro, we decided to fix directly the `material` tag in the urdf stored in the repo.

**Comment by GiulioRomualdi** [comment]
The PR introduces also the `idyntree-model-view-meshcat` application. Thanks to this a model can be easily displayed in meshcat using
```
idyntree-model-view-meshcat -m <model-path>
```

a webpage with the visualizer it will be automatically opened

Since https://github.com/robotology/icub-models-generator/pull/217 has been merged the robots will be displayed as follows

| IRRLICHT | MESHCAT |
|:--------:|:-------:|
| image  | image |
| image | image |

File: bindings/python/scripts/idyntree-model-view-meshcat.py.in [hunk 1]
```diff
@@ -0,0 +1,37 @@
+#!@Python3_EXECUTABLE@
```
Comment (traversaro) [inline bindings/python/scripts/idyntree-model-view-meshcat.py.in:1]: This will hardcode the python executable used at build time in this script. Do you have any specific reason for which you do not want to use the usual Python3 shebang?
```suggestion
#!/usr/bin/env python3
```

**Comment by traversaro** [comment]
Can you update the changelog?

File: bindings/python/scripts/idyntree-model-view-meshcat.py.in (hunk 1 above)
Comment (GiulioRomualdi) [reply bindings/python/scripts/idyntree-model-view-meshcat.py.in:1]: I was thinking if a `virtual enviroment` is used

Comment (traversaro) [reply bindings/python/scripts/idyntree-model-view-meshcat.py.in:1]: For that using `/usr/bin/env python3` should work fine.

Comment (GiulioRomualdi) [reply bindings/python/scripts/idyntree-model-view-meshcat.py.in:1]: Ok! perfect!

Comment (GiulioRomualdi) [reply bindings/python/scripts/idyntree-model-view-meshcat.py.in]: Hi @traversaro, https://github.com/robotology/idyntree/pull/961/commits/3c9e2436f7f2901d1b0516f9b6b6715204949118 should implement what you suggested

**Comment by GiulioRomualdi** [comment]
> Can you update the changelog?

Done in https://github.com/robotology/idyntree/pull/961/commits/69378f6304442dfa7dd40a3a2c30cba6b72aa64b

I opened this PR on `master` and I noticed that the folder structure has been changed in `devel`. How we should proceed? We merge it in master and then me merge master on devel (fixing the conflicts)?

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,6 +6,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Added
+- Implement `idyntree-model-view-meshcat` application. (https://github.com/robotology/idyntree/pull/961)
+### Fixed
+- Once `VisualElement::childElementForName` is called with input equal to `material` a pointer to the `materialInfo` is stored inside the `VisualElement`. (https://github.com/robotology/idyntree/pull/961)
```
Comment (traversaro) [inline CHANGELOG.md:12]:
```suggestion
- Fix loading material and color information from URDF files. (https://github.com/robotology/idyntree/pull/961)
```

Comment (traversaro) [reply CHANGELOG.md:12]: (small comment, what matter to the high level user is the high level effect of the changes, not the low-level details with which really few user may be familiar with)

**Review by traversaro — APPROVED** [review]

**Comment by traversaro** [comment]
@GiulioRomualdi can I merge or you want to do further changes?

**Comment by GiulioRomualdi** [comment]
Feel free to merge it
